### ISSUE TITLE
Bug: `no-throw-literal` false positives with `better-typescript-lib@2.2.0`

### ISSUE THREAD

**Issue by ehoogeveen-medweb** [typescript-eslint/typescript-eslint#6207]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Issue Description

With [better-typescript-lib](https://github.com/uhyo/better-typescript-lib) version `2.2.0`, rule `@typescript-eslint/no-throw-literal` started giving false positives for me. The test file is pretty simple:
```typescript
function loudError(msg: string): Error {
	console.error(msg);
	return new Error(msg);
}

throw loudError('Hello world!');
```
but I've attached a reproduction repo since it requires `better-typescript-lib` to be installed.

With version `2.1.0` of `better-typescript-lib` I don't see the issue, but looking at the [changes](https://github.com/uhyo/better-typescript-lib/compare/v2.1.0...v2.2.0) the only change that looks relevant is the change to `interface Error` where property `cause` went from type `Error` to type `unknown` - and I don't know why that would affect `typescript-eslint`.

I'm not sure if this is a bug in `typescript-eslint` or in `better-typescript-lib`, but I figured there's a better chance of finding the cause here.

cc @uhyo 

### Reproduction Repository Link

https://github.com/ehoogeveen-medweb/typescript-eslint-with-better-typescript-lib

### Repro Steps

1) clone the repo
2) `npm install`
3) `npx eslint test.ts`

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `5.46.1` |
| `@typescript-eslint/parser`            | `5.46.1` |
| `better-typescript-lib`     | `2.2.0` |
| `TypeScript`                           | `4.9.4` |
| `ESLint`                               | `8.29.0` |
| `node`                                 | `18.12.1` |

**Comment by bradzacher** [comment]
We enforce that if you're throwing a type called `Error` that it's the TS library type `Error`, and not some locally defined type called `Error`. It's far from a perfect check because it means you can't ever alias `Error` (but to be fair aliasing a global name like that is smelly code).

If I had to guess - the reason that the rule reports after installing `better-typescript-lib@2.2.0` is 
 because of how TS handles the types the package defines.

The way we check to see if the `Error` type is from TS's types is by asking TS what file the type is defined in, and then asking TS if that file is part of the default library using `program.isSourceFileDefaultLibrary`:

https://github.com/typescript-eslint/typescript-eslint/blob/c3cfe22b4ac152de42a136678dfda50d766a59c8/packages/eslint-plugin/src/rules/no-throw-literal.ts#L69-L77

Having a look at the published version of [`better-typescript-lib`](https://unpkg.com/browse/better-typescript-lib@2.2.0/package.json), the way it works is through some npm behaviour (that I'm surprised is actually allowed! Seems like a massive security hole in NPM TBH that one of your dependencies can alias dependencies in your workspace but that's not related to this issue).

The important bit is that it installs `@typescript/lib-es5`, which ([as of TS 4.5](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-5.html#supporting-lib-from-node_modules)) TS will find and use instead of the inbuilt ES5 libdefs. The es5 libdefs (in case you didn't know) are the ones that contain things like the base `interface Error { ... }` def.

Then I guess that `isSourceFileDefaultLibrary` doesn't return true for the file because essentially its check boils down to `filename.startsWith('node_modules/typescript/lib/')`. IDK if that's a bug or intentional... regardless it's the behaviour!

As for why this only broke with v2.2.0 of their lib? I couldn't say. Would require investigation.

**Comment by ehoogeveen-medweb** [comment]
Thanks for the quick investigation!

Looking in the TypeScript source, function [isSourceFileDefaultLibrary](https://github.com/microsoft/TypeScript/blob/41fa4d7eddfb9691d7f8dcea8741552d5c09c4bf/src/compiler/program.ts#L2557) seems to run function [getDefaultLibFileName](https://github.com/microsoft/TypeScript/blob/41fa4d7eddfb9691d7f8dcea8741552d5c09c4bf/src/compiler/utilitiesPublic.ts#L279) to determine the expected filename.

Maybe `better-typescript-lib` could use these filenames instead of `index.d.ts`? Though I still don't know why it doesn't fail with version `2.1.0`.

**Comment by ehoogeveen-medweb** [comment]
Well, just renaming the file and the reference in its package.json doesn't work, so I guess it won't be that easy.

**Comment by JoshuaKGoldberg** [comment]
Given that this is a relatively rare user need, it _seems_ like there's something `better-typescript-lib` can do, and https://github.com/uhyo/better-typescript-lib/issues/18 is still pending investigation, I'm going to mark this as blocked and take it off our triage queue. But do post back here if you find something that we should change in typescript-eslint! Thanks!

**Comment by uhyo** [comment]
Thank you for reporting this. This is fixed on `better-typescript-lib` side.
Turned out the the fix was as easy as just adding a `/// <reference no-default-lib="true"/>` line to all the library files.
